src/Client/HedgedConnections.cpp

```diff
@@ -366,7 +366,7 @@ int HedgedConnections::getReadyFileDescriptor(AsyncCallback async_callback)
     {
         events_count = epoll.getManyReady(1, &event, false);
         if (!events_count && async_callback)
-            async_callback(epoll.getFileDescriptor(), 0, epoll.getDescription());
+            async_callback(epoll.getFileDescriptor(), settings.receive_timeout, epoll.getDescription());
```

Comment:
But why do we need a receive timeout here? We have a receive timeout for each replica in epoll, so if no replica respond, one of this timeouts will be ready in epoll and we will return from async_callback to process it. Or I missed something?

Reply:
I also think it is ok to use zero here.
As I can see from code, it means `no timeout` for callback. Which is ok, cause timeouts are processed in `HedgedConnections`, and fd from epoll will signal if one is achieved.

Reply:
But who will set it?

With async_socket_for_remote the timeout is set in RemoteQueryExecutorReadContext::setTimer
It uses the timeout that is passed via RemoteQueryExecutorReadContext::setConnectionFD, that called from the async callback.

The timeout that is set in Connection will not work due to non blocking read.

And with use_hedged_requsts and async_socket_for_remote, the async callback will be called from this place with zero timeout, and the timeout will not be set.

Am I missing something?
Yeah the code is complicated.

Also setting it here too should not be a problem in general anyway, it is better to set the timeout one more time instead of hung.

Reply:
> But who will set it?

In HedgedConnections each replica has it's own PacketReceiver, which is responsible for reading packets from socket and process receive timeout. In HedgedConnections we poll file descriptors from these packet receivers, which will be ready if receive timeout has expired or the socket is ready for reading. You can see related code here:
https://github.com/ClickHouse/ClickHouse/blob/e5b96bd93b53d2c1130a249769be1049141ef386/src/Client/PacketReceiver.h#L27
https://github.com/ClickHouse/ClickHouse/blob/e5b96bd93b53d2c1130a249769be1049141ef386/src/Client/PacketReceiver.h#L103
https://github.com/ClickHouse/ClickHouse/blob/e5b96bd93b53d2c1130a249769be1049141ef386/src/Client/PacketReceiver.h#L66

Reply:
>PacketReceiver, which is responsible for reading packets from socket and process receive timeout

@Avogar indeed, thanks!

I believe that the problem that I'm faced with may be related to the lack of patch in my kernel, see #22109